lib/stat_tracker.rb

```diff
@@ -131,6 +131,36 @@ def total_games_per_team_for(season_id)
     end
   end
 
+  def games_for_team(team_id)
+    @games.select {|game| game.away_team_id == team_id || game.home_team_id == team_id}
+  end
+
+  def opponent_by_game_id_for_team(team_id)
+    games_for_team(team_id).reduce({}) do |result, game|
+      result[game.game_id] =
+        if game.away_team_id == team_id 
+          game.home_team_id 
+        else 
+          game.away_team_id
+        end 
+      result
+    end
+  end
+
+  def opponents_and_num_losses_for_team(team_id)
+    result = Hash.new(0)
+    opponent_by_game_id_for_team(team_id).each do |k,v|      
+      @game_teams.each do |game_team|
+        if game_team.game_id == k && game_team.team_id == v && game_team.result == "WIN"
+          result[v] += 1
+        else game_team.game_id == k && game_team.team_id == v && game_team.result != "WIN"
+          result[v] += 0 
+        end
+      end
+    end
+    result
+  end
```

Comment:
Using reduce with hashes is tricky! I thought it would allow us to use three block variables like `|result, key, value|` but it looks like it goes `|result, [key, value]|` which isn't ideal but isn't too bad either. 

I played with this a bit and came up with the following. It adapts your method so that it can be used for calculating both losses and wins based on a `game_result` parameter (which would make it reusable for the `favorite_opponent` method!) and extracts the if/else logic with an additional helper method:

```
def opponents_and_num_losses_for_team(team_id, game_result)
    opponent_by_game_id_for_team(team_id).reduce(Hash.new(0)) do |result, game_id_oppo_ary|
      result[game_id_oppo_ary[1]] += 1 if game_result?(game_result, team_id, game_id_oppo_ary[0])
      result
    end
  end

  def game_result?(result, team_id, game_id)
    game_team = @game_teams.find { |game_team| game_team.game_id == game_id }
    game_team.team_id == team_id && game_team.result == result
  end
```

Reply:
Continuing from slack convo: tried using these methods but was running into issues with game_result? returning true accurately sometimes but false when it should be true other times. I'll keep working on this but wanted to push what I have in case someone else sees something I don't!

Reply:
Ohh interesting! That's why testing is good haha. I'll play with it a bit more and see what's going on. 

Oh also the way I originally set up the `opponents_and_num_losses_for_team` didn't populate a key value pair if there was no match (the hash would not include zero values). Do we need zero values? Because I think we're gonna look for the max wins or max losses so no need for zeroes?

Reply:
I think I found what was happening. The first line of the `game_result?` method is finding the GameTeam object that has a matching game_id but there are always TWO GameTeam objects that have the same game_id. So we need to move the `game_team.team_id == team_id` condition into the #find block

```
def game_result?(result, team_id, game_id)
    game_team = @game_teams.find { |game_team| game_team.game_id == game_id && game_team.team_id == team_id }
    game_team.result == result
  end
```

Also I think the variable names are a bit confusing. game_team is both the local variable and the block variable!

Reply:
Thanks so much for helping me out with this one! It's _finally_ ready for actual review.

Reply:
No worries!! I think this was one of the hardest methods we had to build. You did all the hard work. I just played with the stuff you already had!